### ISSUE TITLE
Strange output when inference is performed on a gpu device other than the cuda default.

### ISSUE THREAD

**Issue by To-jak** [rosinality/stylegan2-pytorch#214]
Hi,

I ran into a strange problem, especially as it does not produce any sort of error / warning message:
On a machine with multiple cuda devices, you cannot obtain a correct output if you perform the inference on a device other than the default one.

I have tested on:
* PyTorch 1.7.1 and CUDA 11.1
* PyTorch 1.4.0 and CUDA 10.1

Here is a minimal code snippet to reproduce the error:

```
import torch
from torchvision import transforms
from model import Generator

def tensor2im(var):
    max = torch.max(var)
    min = torch.min(var)
    var -= min
    var /= max - min
    image = transforms.ToPILImage()(torch.squeeze(var.cpu()))
    return image

def load_generator(path):
    model = Generator(1024, 512, 8, channel_multiplier=2)
    state_dict = torch.load(path, map_location='cpu')
    latent_avg = None if 'latent_avg' not in state_dict else state_dict['latent_avg']
    model.load_state_dict(state_dict['g_ema'])
    model.eval()
    return model, latent_avg

model, latent_avg = load_generator('stylegan2-ffhq-config-f.pt')
model = model.to('cuda:0')

# Using the [512] latent average to do inference, but this work with any latent codes
codes = latent_avg.unsqueeze(0).repeat(18, 1).unsqueeze(0).to('cuda:0')
# Codes.shape: [1, 18, 512]

with torch.no_grad():
    output, _ = model([codes], input_is_latent=True)

output_image = tensor2im(output)
```

With this, you get the correct output for the average latent codes:

![correct_output](https://user-images.githubusercontent.com/33063955/114834305-b7f55880-9dd0-11eb-8a95-090ba4ce3f25.png)

But if you change the `to(cuda:0)` for the model and the latent codes input, you get a totally different looking result:

![wrong_output](https://user-images.githubusercontent.com/33063955/114834451-d9eedb00-9dd0-11eb-8c5b-69ab7a78b21f.png)

This issue seems kind of related to this one: #60 
The quick fix mentioned there is to change `to(cuda:device_number)` with `device = 'cuda'` and then setting the default device to use with `torch.cuda.set_device(device_number)`.

This allows to get the correct result, but it's not the "real solution" as it does not makes possible to load multiple instances of different stylegan2 generators on multiple device at the same time (for example with one instance loaded on `cuda:0` and another one on the `cuda:1` device) and getting correct results for both.

You mentioned in the issue referenced above that it may be related to custom cuda kernel operations. 
At the beginning I thought it may be caused by this kind of lines in the kernels https://github.com/rosinality/stylegan2-pytorch/blob/3dee637b8937bf3830991c066ed8d9cc58afd661/op/upfirdn2d_kernel.cu#L209-L215 but when I modify the index manually to test and compile new kernels it does not seem to change the problem.

Do you have any further idea ?
Thanks!

**Comment by rosinality** [comment]
I found I need to use DeviceGuard for multi device settings. Fixed it at 13f74e3b27aaf50c382addd4c6983dfdb668ab29.

**Comment by To-jak** [comment]
Oh, I see. It works now, thank you for your reactivity!
